### PR TITLE
[ONNX] Fix signed right shift export in the TorchScript exporter

### LINKED ISSUE #191014: torch.onnx.export (TorchScript exporter) gives wrong results for right shift on signed integers
### 🐛 Describe the bug

A right shift on a signed integer tensor exports incorrectly through the legacy TorchScript ONNX exporter (`dynamo=False`). The output is wrong for negative inputs.

In PyTorch, `>>` on a signed integer is an arithmetic shift that rounds toward negative infinity. The exporter lowers it to `Div` by `2**n` (see [symbolic_opset11.py#L1007](https://github.com/pytorch/pytorch/blob/8c48b1e8f8db78e579c11bd36b25ce3299788a6a/torch/onnx/_internal/torchscript_exporter/symbolic_opset11.py#L1007)). ONNX integer `Div` rounds toward zero, so any negative value that shifts out a set bit comes back too large by one.

The code uses ONNX `BitShift` for `UINT8` only. Every signed dtype takes the `Div` path:

```python
@_onnx_symbolic("aten::bitwise_right_shift")
@_onnx_symbolic("aten::__rshift_")
def __rshift_(g, self, other):
    ...
    if JitScalarType.from_value(self) == JitScalarType.UINT8:
        return g.op("BitShift", self, other, direction_s="RIGHT")
    two = g.op("Constant", value_t=torch.tensor(2, dtype=torch.float32))
    if not symbolic_helper._is_fp(self):
        other = g.op("Cast", other, to_i=TensorProtoDataType.FLOAT)
    two_pow = g.op("Pow", two, other)
    two_pow = g.op("Cast", two_pow, to_i=...onnx_type())
    rshift = g.op("Div", self, two_pow)   # rounds toward zero
    return rshift
```

**Repro**

```python
import io

import torch
import onnx
import onnxruntime as ort


class RightShift(torch.nn.Module):
    def forward(self, x):
        return x >> 1


x = torch.tensor([-7, -4, -1, 0, 3, 7], dtype=torch.int8)

buffer = io.BytesIO()
torch.onnx.export(RightShift(), (x,), buffer, dynamo=False, opset_version=17, input_names=["x"])

model = onnx.load_from_string(buffer.getvalue())
ops = [node.op_type for node in model.graph.node]
print("exported ops:", ops)

session = ort.InferenceSession(buffer.getvalue())
onnx_result = session.run(None, {"x": x.numpy()})[0]

print("eager output:", (x >> 1).tolist())
print("onnx output:", onnx_result.tolist())
```

**Output**

```
exported ops: ['Constant', 'Constant', 'Pow', 'Cast', 'Div']
eager output: [-4, -2, -1, 0, 1, 3]
onnx output: [-3, -2, 0, 0, 1, 3]
```

**Expected behavior**

The exported model should match eager, `[-4, -2, -1, 0, 1, 3]`. For signed types the shift must round toward negative infinity, not toward zero.

The `dynamo=True` exporter already gets this right for signed types. It simulates the arithmetic shift with a logical shift plus a sign mask, using a `Less`, `Cast`, `BitShift`, `BitwiseNot`, `BitwiseOr`, `Where` pattern (see `aten_bitwise_right_shift` in onnxscript's torch_lib: https://github.com/microsoft/onnxscript/blob/main/onnxscript/function_libs/torch_lib/ops/core.py#L1474).

**Additional context**

* Affects `aten::__rshift__` and `aten::bitwise_right_shift` for every signed integer dtype (int8, int16, int32, int64).
* Left shift (`Mul` by `2**n`) is fine.
* Source: `torch/onnx/_internal/torchscript_exporter/symbolic_opset11.py`, function `__rshift_`, lines 1007 to 1030, commit `8c48b1e8f8db78e579c11bd36b25ce3299788a6a`.

### Versions

PyTorch version: 2.13.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 19.1.7 (++20250804090312+cd708029e0b2-1~exp1~20250804210325.79)
CMake version: version 4.4.0
Libc version: glibc-2.39

Python version: 3.11.14 (main, Oct 28 2025, 12:11:26) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-6.14.0-37-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 570.172.08
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               AuthenticAMD
Model name:                              AMD Ryzen 9 7950X 16-Core Processor
CPU family:                              25
Model:                                   97
Thread(s) per core:                      2
Core(s) per socket:                      16
Socket(s):                               1
Stepping:                                2
Frequency boost:                         enabled
CPU(s) scaling MHz:                      83%
CPU max MHz:                             5883.0000
CPU min MHz:                             545.0000
BogoMIPS:                                8982.87
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d amd_lbr_pmc_freeze
Virtualization:                          AMD-V
L1d cache:                               512 KiB (16 instances)
L1i cache:                               512 KiB (16 instances)
L2 cache:                                16 MiB (16 instances)
L3 cache:                                64 MiB (2 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-31
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] onnx==1.22.0
[pip3] onnx-ir==0.2.1
[pip3] onnxruntime==1.27.0
[pip3] onnxscript==0.7.1
[pip3] torch==2.13.0+cpu
[pip3] torchao==0.17.0
[conda] Could not collect

cc @justinchuby @titaiwangms

### PR DESCRIPTION
**Pull request by roberto-laudani**
Fixes #191014

### Fix
The exporter lowered signed right shifts to `onnx::Div` by `2**n`, which truncates toward zero, so negative inputs were wrong (e.g. `-7 >> 1` exported as `-3` instead of `-4`). `symbolic_opset9` now computes `(x + m) / 2**n - m`, where `m` is 1 for negative elements and 0 otherwise, rounding toward negative infinity to match PyTorch. int8 and int16 are computed as int32 (the narrowest integer type `onnx::Add` and `onnx::Div` accept before opset 14) and narrowed back losslessly. The unsigned `BitShift` path is unchanged.

cc @justinchuby @titaiwangms

### CODE REVIEW

File: torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py [hunk 1]
```diff
@@ -2174,6 +2199,15 @@ def __rshift_(g: jit_utils.GraphContext, self, other):
     if not symbolic_helper._is_fp(self):
         other = g.op("Cast", other, to_i=_C_onnx.TensorProtoDataType.FLOAT)
     two_pow = g.op("Pow", two, other)
+
+    if self_scalar_type in (
+        _type_utils.JitScalarType.INT8,
```
Comment (Skylion007) [inline torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:2204]: Is there no is integer and is_signed utility here?

Comment (justinchuby) [reply torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:2204]: Probably not. But also this route is deprecated so we don't really invest in improving it.

Comment (roberto-laudani) [reply torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:2204]: @Skylion007 I don't think there is one. `symbolic_helper` defines only [`_is_fp`](https://github.com/pytorch/pytorch/blob/f138aa87c437ad64f49c678ca9c357f33fd20228/torch/onnx/_internal/torchscript_exporter/symbolic_helper.py#L822-L830) and [`_is_bool`](https://github.com/pytorch/pytorch/blob/f138aa87c437ad64f49c678ca9c357f33fd20228/torch/onnx/_internal/torchscript_exporter/symbolic_helper.py#L833), and [`JitScalarType`](https://github.com/pytorch/pytorch/blob/f138aa87c437ad64f49c678ca9c357f33fd20228/torch/onnx/_internal/torchscript_exporter/_type_utils.py#L247-L270) has no integer or signed predicate.

Comment (roberto-laudani) [reply torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:2204]: @justinchuby I am aware the TorchScript exporter is deprecated. The fix is limited to `symbolic_opset9` and `symbolic_opset11` and corrects wrong numerics, `-7 >> 1` currently exports as `-3` instead of `-4`. Would you still take this correction, or would you prefer I close the PR?

Comment (justinchuby) [reply torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:2204]: I am happy to take this PR. Could you complete the EasyCLA step?

Comment (roberto-laudani) [reply torch/onnx/_internal/torchscript_exporter/symbolic_opset9.py:2204]: Done

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
@pytorch merge -i

**Comment by roberto-laudani** [comment]
@justinchuby I think there is a typo: should be pytorchbot
